File: patient-ios-app/MediHubPatient/Core/Models/ClinicalModels.swift
import Foundation

// MARK: - Encounter / Visit Models

struct EncounterDTO: Codable, Identifiable, Hashable {
    let id: String?

    // Backend field names (EncounterResponseDTO.java)
    let encounterDate: String?
    let encounterType: String?
    let staffName: String?
    let departmentName: String?
    let chiefComplaint: String?
    let diagnosisSummary: String?
    let status: String?
    let notes: String?
    let hospitalName: String?
    let appointmentReason: String?

    // Computed aliases used by views
    var date: String? { encounterDate }
    var type: String? { encounterType?.replacingOccurrences(of: "_", with: " ") }
    var providerName: String? { staffName }
    var department: String? { departmentName }
    var reason: String? { chiefComplaint ?? appointmentReason }
    var diagnosis: String? { diagnosisSummary }
    var doctorName: String? { staffName }
}

// MARK: - After Visit Summary

struct AfterVisitSummaryDTO: Codable, Identifiable {
    let id: String?

    // Patient
    let patientName: String?
    let patientMrn: String?

    // Encounter
    let encounterId: String?
    let encounterType: String?

    // Hospital
    let hospitalName: String?

    // Provider
    let dischargingProviderName: String?

    // Dates
    let dischargeDate: String?
    let dischargeTime: String?

    // Disposition / diagnoses
    let disposition: String?
    let dischargeDiagnosis: String?
    let hospitalCourse: String?
    let dischargeCondition: String?

    // Instructions
    let activityRestrictions: String?
    let dietInstructions: String?
    let woundCareInstructions: String?
    let followUpInstructions: String?
    let warningSigns: String?
    let patientEducationProvided: String?

    // Medication reconciliation (keep generic for now)
    let medicationReconciliation: [MedicationReconciliationDTO]?
    let pendingTestResults: [PendingTestResultDTO]?
    let followUpAppointments: [FollowUpAppointmentDTO]?
    let equipmentAndSupplies: [String]?

    // Finalization
    let isFinalized: Bool?
    let finalizedAt: String?
    let additionalNotes: String?

    // Convenience aliases for views
    var providerName: String? { dischargingProviderName }
    var encounterDate: String? { dischargeDate ?? dischargeTime }
    var department: String? { nil }
    var chiefComplaint: String? { hospitalCourse }
    var diagnoses: [String]? {
        guard let d = dischargeDiagnosis, !d.isEmpty else { return nil }
        return [d]
    }
    var treatmentSummary: String? { hospitalCourse }
    var instructions: String? { followUpInstructions }
    var followUpDate: String? {
        followUpAppointments?.first?.appointmentDate
    }
    var medications: [String]? {
        medicationReconciliation?.compactMap {
            [$0.medicationName, $0.dosage, $0.frequency].compactMap { $0 }.joined(separator: " ")
        }
    }
    var status: String? {
        isFinalized == true ? "Finalized" : "Draft"
    }
}

// MARK: - Discharge sub-DTOs

struct MedicationReconciliationDTO: Codable {
    let medicationName: String?
    let dosage: String?
    let frequency: String?
    let route: String?
    let action: String? // CONTINUE, STOP, NEW, MODIFIED
}

struct PendingTestResultDTO: Codable {
    let testName: String?
    let expectedDate: String?
    let notes: String?
}

struct FollowUpAppointmentDTO: Codable {
    let providerName: String?
    let specialty: String?
    let appointmentDate: String?
    let notes: String?
}

// MARK: - Care Team (matches backend CareTeamDTO.java)

struct CareTeamDTO: Codable {
    let primaryCare: PrimaryCareEntry?
    let primaryCareHistory: [PrimaryCareEntry]?

    /// Flatten into displayable members list
    var allMembers: [PrimaryCareEntry] {
        var list: [PrimaryCareEntry] = []
        if let pc = primaryCare { list.append(pc) }
        if let history = primaryCareHistory {
            list.append(contentsOf: history.filter { !$0.current })
        }
        return list
    }
}

struct PrimaryCareEntry: Codable, Identifiable {
    let id: String?
    let hospitalId: String?
    let hospitalName: String?
    let doctorUserId: String?
    let doctorDisplay: String?
    let startDate: String?
    let endDate: String?
    let current: Bool

    // Convenience for views
    var displayName: String { doctorDisplay ?? "Provider" }
}

// Legacy — kept so CareTeamMemberRow still compiles if referenced
struct CareTeamMemberDTO: Codable, Identifiable {
    var id: String? { name }
    let name: String?
    let role: String?
    let specialty: String?
    let phone: String?
    let email: String?
    let isPrimary: Bool?
}

// MARK: - Document Models

struct DocumentDTO: Codable, Identifiable {
    let id: String?
    let fileName: String?
    let fileType: String?
    let fileSize: Int?
    let uploadedAt: String?
    let description: String?
    let category: String?
    let uploadedBy: String?
    let downloadUrl: String?
}

// MARK: - Notification Models

struct NotificationDTO: Codable, Identifiable {
    let id: String?
    let title: String?
    let message: String?
    let type: String?
    let read: Bool?
    let createdAt: String?
    let actionUrl: String?

    var isRead: Bool {
        read ?? false
    }
}

// MARK: - Chat / Message Models

struct ChatThreadDTO: Codable, Identifiable, Hashable {
    let id: String?
    let recipientName: String?
    let recipientRole: String?
    let lastMessage: String?
    let lastMessageAt: String?
    let unreadCount: Int?
}

struct ChatMessageDTO: Codable, Identifiable {
    let id: String?
    let senderId: String?
    let senderName: String?
    let content: String?
    let sentAt: String?
    let read: Bool?
    let attachmentUrl: String?
}

struct SendMessageRequest: Encodable {
    let content: String
    let attachmentUrl: String?
}

// MARK: - Referral Models

struct ReferralDTO: Codable, Identifiable {
    let id: String?
    let referralDate: String?
    let status: String?
    let reason: String?
    let fromDoctorName: String?
    let toDoctorName: String?
    let toSpecialty: String?
    let toHospitalName: String?
    let notes: String?
    let urgency: String?
}

// MARK: - Treatment Plan Models

struct TreatmentPlanDTO: Codable, Identifiable {
    let id: String?
    let title: String?
    let startDate: String?
    let endDate: String?
    let status: String?
    let goals: String?
    let interventions: String?
    let doctorName: String?
    let notes: String?
}

// MARK: - Consent Models

struct ConsentDTO: Codable, Identifiable {
    let id: String?
    let fromHospitalId: String?
    let fromHospitalName: String?
    let toHospitalId: String?
    let toHospitalName: String?
    let consentType: String?
    let purpose: String?
    let grantedAt: String?
    let expiresAt: String?
    let status: String?
}

struct GrantConsentRequest: Encodable {
    let fromHospitalId: String
    let toHospitalId: String
    let purpose: String?
    let consentExpiration: String?
}

// MARK: - Access Log Models

struct AccessLogDTO: Codable, Identifiable {
    let id: String?
    let accessedBy: String?
    let accessedByRole: String?
    let accessType: String?
    let resourceAccessed: String?
    let accessedAt: String?
    let ipAddress: String?
}

// MARK: - Immunization Models

struct ImmunizationDTO: Codable, Identifiable {
    let id: String?
    let vaccineName: String?
    let dateAdministered: String?
    let provider: String?
    let status: String?
}

// MARK: - Proxy / Family Access Models

struct ProxyResponse: Codable, Identifiable {
    let id: String?
    let grantorPatientId: String?
    let grantorName: String?
    let proxyUserId: String?
    let proxyUsername: String?
    let proxyDisplayName: String?
    let relationship: String?
    let status: String?
    let permissions: String?
    let expiresAt: String?
    let revokedAt: String?
    let notes: String?
    let createdAt: String?

    /// Convenience: split comma-separated permissions into an array for UI display
    var permissionsList: [String] {
        permissions?.split(separator: ",").map { String($0).trimmingCharacters(in: .whitespaces) } ?? []
    }
}

struct ProxyGrantRequest: Encodable {
    let proxyUsername: String
    let relationship: String
    let permissions: String // comma-separated: "VIEW_RECORDS,VIEW_APPOINTMENTS"
    let expiresAt: String?
    let notes: String?
}

// MARK: - Patient Payment

struct PatientPaymentRequest: Encodable {
    let amount: Double
    let paymentMethod: String
    let transactionReference: String?
    let notes: String?
}

File: patient-android-app/app/src/main/java/com/bitnesttechs/hms/patient/core/models/ClinicalModels.kt
package com.bitnesttechs.hms.patient.core.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class EncounterDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "encounterDate") val encounterDate: String = "",
    @Json(name = "type") val type: String? = null,
    @Json(name = "encounterType") val encounterType: String = "",
    @Json(name = "status") val status: String = "",
    @Json(name = "chiefComplaint") val chiefComplaint: String? = null,
    @Json(name = "diagnosis") val diagnosis: String? = null,
    @Json(name = "doctorName") val doctorName: String? = null,
    @Json(name = "department") val department: String? = null,
    @Json(name = "departmentName") val departmentName: String? = null,
    @Json(name = "hospitalName") val hospitalName: String? = null,
    @Json(name = "notes") val notes: String? = null
)

@JsonClass(generateAdapter = true)
data class DischargeSummaryDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "encounterId") val encounterId: String? = null,
    @Json(name = "admissionDate") val admissionDate: String? = null,
    @Json(name = "dischargeDate") val dischargeDate: String? = null,
    @Json(name = "primaryDiagnosis") val primaryDiagnosis: String? = null,
    @Json(name = "instructions") val instructions: String? = null,
    @Json(name = "dischargeInstructions") val dischargeInstructions: String? = null,
    @Json(name = "followUpDate") val followUpDate: String? = null,
    @Json(name = "followUpInstructions") val followUpInstructions: String? = null,
    @Json(name = "doctorName") val doctorName: String? = null,
    @Json(name = "departmentName") val departmentName: String? = null
)

@JsonClass(generateAdapter = true)
data class CareTeamDto(
    @Json(name = "primaryPhysician") val primaryPhysician: CareTeamMemberDto? = null,
    @Json(name = "members") val members: List<CareTeamMemberDto> = emptyList()
)

@JsonClass(generateAdapter = true)
data class CareTeamMemberDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "name") val name: String = "",
    @Json(name = "role") val role: String? = null,
    @Json(name = "specialty") val specialty: String? = null,
    @Json(name = "phone") val phone: String? = null,
    @Json(name = "email") val email: String? = null,
    @Json(name = "department") val department: String? = null
)

@JsonClass(generateAdapter = true)
data class DocumentDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "name") val name: String = "",
    @Json(name = "type") val type: String? = null,
    @Json(name = "fileType") val fileType: String? = null,
    @Json(name = "documentType") val documentType: String? = null,
    @Json(name = "uploadedAt") val uploadedAt: String = "",
    @Json(name = "fileUrl") val fileUrl: String? = null,
    @Json(name = "fileSize") val fileSize: Long? = null,
    @Json(name = "description") val description: String? = null
)

@JsonClass(generateAdapter = true)
data class NotificationDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "title") val title: String = "",
    @Json(name = "message") val message: String = "",
    @Json(name = "type") val type: String? = null,
    @Json(name = "isRead") val isRead: Boolean = false,
    @Json(name = "createdAt") val createdAt: String = "",
    @Json(name = "actionUrl") val actionUrl: String? = null
)

@JsonClass(generateAdapter = true)
data class ChatThreadDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "subject") val subject: String = "",
    @Json(name = "participantName") val participantName: String? = null,
    @Json(name = "lastMessage") val lastMessage: String? = null,
    @Json(name = "lastMessageAt") val lastMessageAt: String? = null,
    @Json(name = "unreadCount") val unreadCount: Int = 0
)

@JsonClass(generateAdapter = true)
data class ChatMessageDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "threadId") val threadId: String = "",
    @Json(name = "content") val content: String = "",
    @Json(name = "senderName") val senderName: String? = null,
    @Json(name = "sentAt") val sentAt: String = "",
    @Json(name = "isFromPatient") val isFromPatient: Boolean = false,
    @Json(name = "isRead") val isRead: Boolean = false
)

@JsonClass(generateAdapter = true)
data class SendMessageRequest(
    @Json(name = "content") val content: String,
    @Json(name = "subject") val subject: String? = null
)

@JsonClass(generateAdapter = true)
data class ReferralDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "referralDate") val referralDate: String? = null,
    @Json(name = "referralType") val referralType: String? = null,
    @Json(name = "referredTo") val referredTo: String? = null,
    @Json(name = "specialistName") val specialistName: String? = null,
    @Json(name = "specialty") val specialty: String? = null,
    @Json(name = "reason") val reason: String? = null,
    @Json(name = "status") val status: String = "",
    @Json(name = "notes") val notes: String? = null
)

@JsonClass(generateAdapter = true)
data class TreatmentPlanDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "title") val title: String = "",
    @Json(name = "description") val description: String? = null,
    @Json(name = "startDate") val startDate: String? = null,
    @Json(name = "endDate") val endDate: String? = null,
    @Json(name = "status") val status: String = "",
    @Json(name = "goals") val goals: List<String> = emptyList(),
    @Json(name = "createdBy") val createdBy: String? = null
)

@JsonClass(generateAdapter = true)
data class ConsentDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "type") val type: String = "",
    @Json(name = "consentType") val consentType: String = "",
    @Json(name = "title") val title: String = "",
    @Json(name = "description") val description: String? = null,
    @Json(name = "status") val status: String = "",
    @Json(name = "isGranted") val isGranted: Boolean = false,
    @Json(name = "recipientName") val recipientName: String? = null,
    @Json(name = "grantedAt") val grantedAt: String? = null,
    @Json(name = "expiresAt") val expiresAt: String? = null
)

@JsonClass(generateAdapter = true)
data class GrantConsentRequest(
    @Json(name = "granted") val granted: Boolean,
    @Json(name = "notes") val notes: String? = null
)

@JsonClass(generateAdapter = true)
data class AccessLogDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "accessedBy") val accessedBy: String = "",
    @Json(name = "accessedAt") val accessedAt: String = "",
    @Json(name = "action") val action: String = "",
    @Json(name = "resourceType") val resourceType: String? = null,
    @Json(name = "ipAddress") val ipAddress: String? = null
)

@JsonClass(generateAdapter = true)
data class ImmunizationDto(
    @Json(name = "id") val id: String = "",
    @Json(name = "vaccineName") val vaccineName: String = "",
    @Json(name = "administeredDate") val administeredDate: String = "",
    @Json(name = "administeredBy") val administeredBy: String? = null,
    @Json(name = "manufacturer") val manufacturer: String? = null,
    @Json(name = "lotNumber") val lotNumber: String? = null,
    @Json(name = "nextDoseDate") val nextDoseDate: String? = null,
    @Json(name = "nextDueDate") val nextDueDate: String? = null,
    @Json(name = "notes") val notes: String? = null
)
